cordyceps/src/sorted_list.rs

```diff
@@ -0,0 +1,459 @@
+//! [Intrusive], singly-linked, sorted, linked-list.
+//!
+//! See the documentation for the [`SortedList`] and [`SortedListIter`] types for
+//! details.
+//!
+//! [intrusive]: crate#intrusive-data-structures
+
+use crate::{Linked, Stack};
+use core::{marker::PhantomData, ptr::NonNull};
+
+pub use crate::stack::Links;
+
+/// A sorted singly linked list
+///
+/// This behaves similar to [`Stack`], in that it is a singly linked list,
+/// however items are stored in an ordered fashion. This means that insertion
+/// is an `O(n)` operation, and retrieval of the first item is an `O(1)` operation.
+///
+/// It allows for a user selected ordering operation. If your type `T` implements
+/// [`Ord`]:
+///
+/// * Consider using [`SortedList::new_min()`] if you want **smallest** items sorted first.
+/// * Consider using [`SortedList::new_max()`] if you want **largest** items sorted first.
+///
+/// If your type `T` does NOT implement [`Ord`], or you want to use
+/// a custom sorting anyway, consider using [`SortedList::new_custom()`]
+#[derive(Debug)]
+pub struct SortedList<T: Linked<Links<T>>> {
+    head: Option<NonNull<T>>,
+    // Returns if LHS is less/same/greater than RHS
+    func: fn(&T, &T) -> core::cmp::Ordering,
+}
+
+#[inline]
+fn invert_sort<T: core::cmp::Ord>(a: &T, b: &T) -> core::cmp::Ordering {
+    // Inverted sort order!
+    T::cmp(b, a)
+}
+
+impl<T> SortedList<T>
+where
+    T: Linked<Links<T>>,
+    T: core::cmp::Ord,
+{
+    /// Create a new (empty) sorted list, sorted LEAST FIRST
+    ///
+    /// * Consider using [`SortedList::new_max()`] if you want **largest** items sorted first.
+    /// * Consider using [`SortedList::new_custom()`] if you want to provide your own sorting
+    ///   implementation.
+    ///
+    /// If two items are considered of equal value, new values will be placed AFTER
+    /// old values.
+    pub const fn new_min() -> Self {
+        Self::new_custom(T::cmp)
+    }
+
+    /// Create a new sorted list, consuming the stack, sorted LEAST FIRST
+    pub fn from_stack_min(stack: Stack<T>) -> Self {
+        Self::from_stack_custom(stack, T::cmp)
+    }
+
+    /// Create a new (empty) sorted list, sorted GREATEST FIRST
+    ///
+    /// * Consider using [`SortedList::new_min()`] if you want **smallest** items sorted first.
+    /// * Consider using [`SortedList::new_custom()`] if you want to provide your own sorting
+    ///   implementation.
+    ///
+    /// If two items are considered of equal value, new values will be placed AFTER
+    /// old values.
+    pub const fn new_max() -> Self {
+        Self::new_custom(invert_sort::<T>)
+    }
+
+    /// Create a new sorted list, consuming the stack, sorted GREATEST FIRST
```

Comment:
It would be nice if the docs for these had examples showing the min- and max-ordering behaviors...but, not a blocker for this PR if that's too much effort, we can add them later.